### DISCUSSION TITLE
`select(model).from_statement(...)` has `is_select = False`

### DISCUSSION THREAD

**Discussion by samscott89** [sqlalchemy/sqlalchemy#11219] [Usage Questions] [answered]
Is it intentional that `is_select` return `False` when using `from_statement` in the following? If so, is there a suggested way to in the `do_orm_execute` code to determine whether the query is a select?

Example code: 

```py
from sqlalchemy.orm import declarative_base, sessionmaker
from sqlalchemy import Column, String, event, select, text, create_engine

engine = create_engine("sqlite:///:memory:", echo=True, future=True)
Session = sessionmaker(engine, expire_on_commit=False)

Base = declarative_base()


class Foo(Base):
    __tablename__ = "foo"

    name = Column(String, primary_key=True)


Base.metadata.create_all(engine)

was_select = None

@event.listens_for(Session, "do_orm_execute")
def do_orm_execute(execute_state):
    global was_select
    was_select = execute_state.is_select
    print(execute_state.is_select)

session = Session()
with Session() as session:
    with session.begin():
        query = select(Foo).filter_by(name="bar")
        session.execute(query)
        assert was_select == True # succeeds
        query = select(Foo).from_statement(text("select * from foo where name = 'bar'"))
        session.execute(query)
        assert was_select == True, "expected was_select to be True"
```

Edit to add: on digging a little more, it's pretty clear the root of this is that `FromStatement.is_select` returns `False`: 

```
(Pdb) p execute_state.statement
<sqlalchemy.orm.context.FromStatement object at 0x104442e70>
(Pdb) p execute_state.statement.is_select
False
```

Updated the title to reflect what I think is the main problem. I'll keep the code as-is though, because it shows the context of what I'm trying to achieve.

Note: this is specific to sqlalchemy 2.0, on 1.4 the above returns `True` for `is_select

**Comment by zzzeek** [comment, marked as answer]
The problem here is that this is not a SELECT at the SQL level:

```
stmt = select(User).from_statement(insert(User).values(name="some name").returning(User))
```

that emits only INSERT..RETURNING.

what do you think `is_select` should be in that case?    less controversial is a new attribute `is_from_statement`

**Comment by zzzeek** [comment]
a patch for this is linked from #11220

**Comment by samscott89** [comment]
Thank you!  I agree that we would _not_ want the provided example  to `is_select = True` and the proposed patch looks perfect.

**Reply by zzzeek** [reply]
is_select will work in the ~~proposed~~ merged patch as long as your from_statement() is itself against a Select

**Reply by samscott89** [reply]
Yep that makes sense. Sorry by "provided example" I meant the "insert .. returning" example you provided, not my original example.
